internal/jimm/offer/authorizer.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025 Canonical.
+
+package offer
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/names/v5"
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+type offerAuthorizer struct {
+	store   *db.Database
+	authSvc *openfga.OFGAClient
+}
+
+// NewOfferAuthorizer returns a new OfferAuthorizer that provides methods to
+// check if a user is a consumer of an application offer.
+func NewOfferAuthorizer(store *db.Database, authSvc *openfga.OFGAClient) (*offerAuthorizer, error) {
+	if store == nil {
+		return nil, errors.E("group store cannot be nil")
+	}
+	if authSvc == nil {
+		return nil, errors.E("group authorisation service cannot be nil")
+	}
+	return &offerAuthorizer{store, authSvc}, nil
+}
+
+// IsUserConsumerForOffer checks if a user is a consumer of an application offer.
+// If the user is local, it is mapped to its corresponding external user
+// based on the model migration user mapping.
+func (offerAuth *offerAuthorizer) IsUserConsumerForOffer(ctx context.Context, userTag names.UserTag, offerTag names.ApplicationOfferTag) (bool, error) {
+	var userIdentifier string
+	var err error
+	if userTag.IsLocal() {
+		userIdentifier, err = offerAuth.resolveLocalUserToExternalUser(ctx, userTag.Id(), offerTag)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		userIdentifier = userTag.Id()
+	}
+	identity, err := dbmodel.NewIdentity(userIdentifier)
+	if err != nil {
+		return false, err
+	}
+	user := openfga.NewUser(
+		identity,
+		offerAuth.authSvc,
+	)
+	return user.IsApplicationOfferConsumer(ctx, offerTag)
+}
+
+func (offerAuth *offerAuthorizer) resolveLocalUserToExternalUser(ctx context.Context, localUsername string, offerTag names.ApplicationOfferTag) (string, error) {
+	offer := dbmodel.ApplicationOffer{
+		UUID: offerTag.Id(),
+	}
+	err := offerAuth.store.GetApplicationOffer(ctx, &offer)
+	if err != nil {
+		return "", err
+	}
+	incomingMigration := dbmodel.IncomingModelMigration{
+		ModelUUID: sql.NullString{
+			String: offer.Model.UUID.String,
+			Valid:  true,
+		},
+	}
+	err = offerAuth.store.GetIncomingModelMigration(ctx, &incomingMigration)
```

Comment:
Don't use this table, as it indicates a model migration has not yet completed. Use the user mapping table.